File: cpp/equity-var-engine/tests/test_historical.cpp
// Historical-simulation VaR: hand-exact quantiles on tiny arrays, BRW
// weights, EWMA filtering, sqrt-time scaling, input validation.
#include <gtest/gtest.h>

#include <cmath>
#include <numeric>
#include <vector>

#include "eqvar/historical.hpp"

using namespace eqvar;

namespace {

// 100-day deterministic P&L series with a fat left tail.
std::vector<double> demo_pnl() {
    std::vector<double> pnl(100);
    for (int t = 0; t < 100; ++t) {
        pnl[t] = 100.0 * std::sin(0.7 * t + 1.0) - (t % 17 == 0 ? 150.0 : 0.0);
    }
    return pnl;
}

}  // namespace

TEST(QuantileLinear, HandExactTinyArrays) {
    // NumPy type-7: h = (n-1) q, linear interpolation between order stats.
    const std::vector<double> x = {1.0, 2.0, 3.0, 4.0};
    EXPECT_DOUBLE_EQ(quantile_linear(x, 0.0), 1.0);
    EXPECT_DOUBLE_EQ(quantile_linear(x, 1.0), 4.0);
    EXPECT_DOUBLE_EQ(quantile_linear(x, 0.5), 2.5);
    EXPECT_DOUBLE_EQ(quantile_linear(x, 0.25), 1.75);  // h = 0.75
    const std::vector<double> unsorted = {3.0, 1.0, 2.0};
    EXPECT_DOUBLE_EQ(quantile_linear(unsorted, 0.5), 2.0);
    EXPECT_DOUBLE_EQ(quantile_linear(unsorted, 0.75), 2.5);  // h = 1.5
    const std::vector<double> single = {7.0};
    EXPECT_DOUBLE_EQ(quantile_linear(single, 0.3), 7.0);
}

TEST(QuantileLinear, Validation) {
    EXPECT_THROW(quantile_linear({}, 0.5), std::invalid_argument);
    const std::vector<double> x = {1.0, 2.0};
    EXPECT_THROW(quantile_linear(x, -0.1), std::invalid_argument);
    EXPECT_THROW(quantile_linear(x, 1.1), std::invalid_argument);
}

TEST(HistoricalVar, MatchesQuantileAndIsMonotoneInAlpha) {
    const std::vector<double> pnl = demo_pnl();
    EXPECT_DOUBLE_EQ(historical_var(pnl, 0.05), -quantile_linear(pnl, 0.05));
    // Deeper tail => larger VaR.
    EXPECT_GE(historical_var(pnl, 0.01), historical_var(pnl, 0.05));
    EXPECT_GE(historical_var(pnl, 0.05), historical_var(pnl, 0.10));
}

TEST(HistoricalVar, Validation) {
    std::vector<double> tiny(kMinHistObs - 1, 0.0);
    EXPECT_THROW(historical_var(tiny, 0.01), std::invalid_argument);
    const std::vector<double> pnl = demo_pnl();
    EXPECT_THROW(historical_var(pnl, 0.0), std::invalid_argument);
    EXPECT_THROW(historical_var(pnl, 0.5), std::invalid_argument);
    std::vector<double> bad = pnl;
    bad[10] = std::nan("");
    EXPECT_THROW(historical_var(bad, 0.01), std::invalid_argument);
}

TEST(BrwWeights, SumToOneAndMonotoneInRecency) {
    for (std::size_t n : {std::size_t{1}, std::size_t{5}, std::size_t{250}}) {
        const std::vector<double> w = brw_weights(n, 0.98);
        ASSERT_EQ(w.size(), n);
        const double sum = std::accumulate(w.begin(), w.end(), 0.0);
        EXPECT_NEAR(sum, 1.0, 1e-12) << "n=" << n;
        for (std::size_t i = 1; i < n; ++i) {
            EXPECT_GT(w[i], w[i - 1]) << "weights must increase with recency";
        }
    }
    // Exact ratio between consecutive weights is 1/lam.
    const std::vector<double> w = brw_weights(10, 0.95);
    EXPECT_NEAR(w[9] / w[8], 1.0 / 0.95, 1e-12);
    EXPECT_THROW(brw_weights(10, 1.0), std::invalid_argument);
    EXPECT_THROW(brw_weights(0, 0.98), std::invalid_argument);
}

TEST(AgeWeightedVar, LambdaNearOneRecoversStepHistorical) {
    const std::vector<double> pnl = demo_pnl();
    // lam -> 1: BRW converges to the equal-weight step-CDF inversion.  Use an
    // alpha safely between the k/n weight boundaries (0.05, 0.06) so the tiny
    // residual age tilt cannot flip the selected order statistic.
    const double brw = age_weighted_var(pnl, 0.053, 0.999999);
    std::vector<double> sorted = pnl;
    std::sort(sorted.begin(), sorted.end());
    // Step inversion at alpha = 0.053 with n = 100 picks the 6th order stat.
    EXPECT_DOUBLE_EQ(brw, -sorted[5]);
}

TEST(AgeWeightedVar, RecentCrashDominatesUnderLowLambda) {
    // A large loss on the most recent day must raise BRW VaR far more than
    // the same loss buried at the start of the window.
    std::vector<double> recent(100, 10.0), old(100, 10.0);
    recent.back() = -500.0;
    old.front() = -500.0;
    const double var_recent = age_weighted_var(recent, 0.05, 0.94);
    const double var_old = age_weighted_var(old, 0.05, 0.94);
    EXPECT_GT(var_recent, var_old);
    EXPECT_DOUBLE_EQ(var_recent, 500.0);  // crash weight (1-l)/(1-l^n) > 5 %
}

TEST(EwmaVolatility, NoLookAheadAndRiskMetricsRecursion) {
    const std::vector<double> x = {1.0, -2.0, 3.0, -1.0, 2.0};
    const double lam = 0.94;
    const std::vector<double> sig = ewma_volatility(x, lam);
    ASSERT_EQ(sig.size(), x.size());
    // Seed = population variance; recursion uses x[t-1] only (no look-ahead).
    double mu = 0.0;
    for (double v : x) mu += v;
    mu /= 5.0;
    double s2 = 0.0;
    for (double v : x) s2 += (v - mu) * (v - mu);
    s2 /= 5.0;
    EXPECT_NEAR(sig[0], std::sqrt(s2), 1e-14);
    for (std::size_t t = 1; t < x.size(); ++t) {
        s2 = lam * s2 + (1.0 - lam) * x[t - 1] * x[t - 1];
        EXPECT_NEAR(sig[t], std::sqrt(s2), 1e-14) << "t=" << t;
    }
    EXPECT_THROW(ewma_volatility(std::vector<double>{1.0}, lam), std::invalid_argument);
    EXPECT_THROW(ewma_volatility(x, 0.0), std::invalid_argument);
}

TEST(FilteredHistoricalVar, ScaleEquivariantAndRegimeResponsive) {
    const std::vector<double> pnl = demo_pnl();
    // Positive homogeneity: FHS(c pnl) = c FHS(pnl).
    std::vector<double> scaled = pnl;
    for (double& v : scaled) v *= 3.0;
    EXPECT_NEAR(filtered_historical_var(scaled, 0.05), 3.0 * filtered_historical_var(pnl, 0.05),
                1e-9);
    // A quiet recent regime must cut FHS VaR below plain historical VaR.
    std::vector<double> calm = pnl;
    for (std::size_t t = 60; t < calm.size(); ++t) calm[t] *= 0.05;
    EXPECT_LT(filtered_historical_var(calm, 0.05), historical_var(calm, 0.05));
}

TEST(SqrtTime, ScalingAndValidation) {
    EXPECT_DOUBLE_EQ(scale_var_sqrt_time(100.0, 1), 100.0);
    EXPECT_DOUBLE_EQ(scale_var_sqrt_time(100.0, 4), 200.0);
    EXPECT_NEAR(scale_var_sqrt_time(50.0, 10), 50.0 * std::sqrt(10.0), 1e-12);
    EXPECT_THROW(scale_var_sqrt_time(100.0, 0), std::invalid_argument);
}

File: rust/equity-var-engine/tests/test_historical.rs
//! Historical-simulation VaR: hand-exact quantiles on tiny arrays, BRW
//! weights, EWMA filtering, sqrt-time scaling, input validation.

use eq_var_engine::historical::*;

fn demo_pnl() -> Vec<f64> {
    (0..100)
        .map(|t| {
            let td = t as f64;
            100.0 * (0.7 * td + 1.0).sin() - if t % 17 == 0 { 150.0 } else { 0.0 }
        })
        .collect()
}

#[test]
fn quantile_linear_hand_exact_tiny_arrays() {
    let x = [1.0, 2.0, 3.0, 4.0];
    assert!((linear_quantile(&x, 0.0).unwrap() - 1.0).abs() < 1e-15);
    assert!((linear_quantile(&x, 1.0).unwrap() - 4.0).abs() < 1e-15);
    assert!((linear_quantile(&x, 0.5).unwrap() - 2.5).abs() < 1e-15);
    assert!((linear_quantile(&x, 0.25).unwrap() - 1.75).abs() < 1e-15); // h = 0.75
    let unsorted = [3.0, 1.0, 2.0];
    assert!((linear_quantile(&unsorted, 0.5).unwrap() - 2.0).abs() < 1e-15);
    assert!((linear_quantile(&unsorted, 0.75).unwrap() - 2.5).abs() < 1e-15); // h = 1.5
    let single = [7.0];
    assert!((linear_quantile(&single, 0.3).unwrap() - 7.0).abs() < 1e-15);
}

#[test]
fn quantile_linear_validation() {
    assert!(linear_quantile(&[], 0.5).is_err());
    let x = [1.0, 2.0];
    assert!(linear_quantile(&x, -0.1).is_err());
    assert!(linear_quantile(&x, 1.1).is_err());
}

#[test]
fn historical_var_matches_quantile_and_monotone_in_alpha() {
    let pnl = demo_pnl();
    assert!(
        (historical_var(&pnl, 0.05).unwrap() - (-linear_quantile(&pnl, 0.05).unwrap())).abs()
            < 1e-12
    );
    assert!(historical_var(&pnl, 0.01).unwrap() >= historical_var(&pnl, 0.05).unwrap());
    assert!(historical_var(&pnl, 0.05).unwrap() >= historical_var(&pnl, 0.10).unwrap());
}

#[test]
fn historical_var_validation() {
    let tiny = vec![0.0; MIN_OBS - 1];
    assert!(historical_var(&tiny, 0.01).is_err());
    let pnl = demo_pnl();
    assert!(historical_var(&pnl, 0.0).is_err());
    assert!(historical_var(&pnl, 0.5).is_err());
    let mut bad = pnl.clone();
    bad[10] = f64::NAN;
    assert!(historical_var(&bad, 0.01).is_err());
}

#[test]
fn brw_weights_sum_to_one_and_monotone_in_recency() {
    for n in [1usize, 5, 250] {
        let w = brw_weights(n, 0.98).unwrap();
        assert_eq!(w.len(), n);
        let sum: f64 = w.iter().sum();
        assert!((sum - 1.0).abs() < 1e-12, "n={n}");
        for i in 1..n {
            assert!(w[i] > w[i - 1], "weights must increase with recency");
        }
    }
    let w = brw_weights(10, 0.95).unwrap();
    assert!((w[9] / w[8] - 1.0 / 0.95).abs() < 1e-12);
    assert!(brw_weights(10, 1.0).is_err());
    assert!(brw_weights(0, 0.98).is_err());
}

#[test]
fn age_weighted_var_lambda_near_one_recovers_step_historical() {
    let pnl = demo_pnl();
    let brw = age_weighted_var(&pnl, 0.053, 0.999999).unwrap();
    let mut sorted = pnl.clone();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    assert!((brw - (-sorted[5])).abs() < 1e-9);
}

#[test]
fn age_weighted_var_recent_crash_dominates_under_low_lambda() {
    let mut recent = vec![10.0; 100];
    let mut old = vec![10.0; 100];
    *recent.last_mut().unwrap() = -500.0;
    old[0] = -500.0;
    let var_recent = age_weighted_var(&recent, 0.05, 0.94).unwrap();
    let var_old = age_weighted_var(&old, 0.05, 0.94).unwrap();
    assert!(var_recent > var_old);
    assert!((var_recent - 500.0).abs() < 1e-9);
}

#[test]
fn ewma_volatility_no_lookahead_and_riskmetrics_recursion() {
    let x = [1.0, -2.0, 3.0, -1.0, 2.0];
    let lam = 0.94;
    let sig = ewma_volatility(&x, lam).unwrap();
    assert_eq!(sig.len(), x.len());
    let mu: f64 = x.iter().sum::<f64>() / 5.0;
    let mut s2: f64 = x.iter().map(|v| (v - mu) * (v - mu)).sum::<f64>() / 5.0;
    assert!((sig[0] - s2.sqrt()).abs() < 1e-14);
    for t in 1..x.len() {
        s2 = lam * s2 + (1.0 - lam) * x[t - 1] * x[t - 1];
        assert!((sig[t] - s2.sqrt()).abs() < 1e-14, "t={t}");
    }
    assert!(ewma_volatility(&[1.0], lam).is_err());
    assert!(ewma_volatility(&x, 0.0).is_err());
}

#[test]
fn filtered_historical_var_scale_equivariant_and_regime_responsive() {
    let pnl = demo_pnl();
    let scaled: Vec<f64> = pnl.iter().map(|v| v * 3.0).collect();
    assert!(
        (filtered_historical_var(&scaled, 0.05, 0.94).unwrap()
            - 3.0 * filtered_historical_var(&pnl, 0.05, 0.94).unwrap())
        .abs()
            < 1e-9
    );
    let mut calm = pnl.clone();
    for v in calm.iter_mut().skip(60) {
        *v *= 0.05;
    }
    assert!(
        filtered_historical_var(&calm, 0.05, 0.94).unwrap()
            < historical_var(&calm, 0.05).unwrap()
    );
}

#[test]
fn sqrt_time_scaling_and_validation() {
    assert!((scale_var_sqrt_time(100.0, 1).unwrap() - 100.0).abs() < 1e-15);
    assert!((scale_var_sqrt_time(100.0, 4).unwrap() - 200.0).abs() < 1e-12);
    assert!((scale_var_sqrt_time(50.0, 10).unwrap() - 50.0 * 10f64.sqrt()).abs() < 1e-12);
    assert!(scale_var_sqrt_time(100.0, 0).is_err());
}

#[test]
fn overlapping_horizon_pnl_basic() {
    let pnl = demo_pnl();
    let windows = overlapping_horizon_pnl(&pnl, 5).unwrap();
    assert_eq!(windows.len(), pnl.len() - 5 + 1);
    let hand: f64 = pnl[0..5].iter().sum();
    assert!((windows[0] - hand).abs() < 1e-9);
    assert!(overlapping_horizon_pnl(&pnl, 0).is_err());
}
